include/net/mqtt_sn.h

```diff
@@ -0,0 +1,197 @@
+	MQTT_SN_QOS_0, /**< QOS 0 */
+	MQTT_SN_QOS_1, /**< QOS 1 */
+	MQTT_SN_QOS_2, /**< QOS 2 */
+	MQTT_SN_QOS_M1 /**< QOS -1 */
+};
+
+enum __attribute__((packed)) mqtt_sn_topic_type {
+	MQTT_SN_TOPIC_TYPE_NORMAL,
+	MQTT_SN_TOPIC_TYPE_PREDEF,
+	MQTT_SN_TOPIC_TYPE_SHORT
+};
+
+enum __attribute__((packed)) mqtt_sn_return_code {
+	MQTT_SN_CODE_ACCEPTED = 0x00, /**< Accepted */
+	MQTT_SN_CODE_REJECTED_CONGESTION = 0x01, /**< Rejected: congestion */
+	MQTT_SN_CODE_REJECTED_TOPIC_ID = 0x02, /**< Rejected: Invalid Topic ID */
+	MQTT_SN_CODE_REJECTED_NOTSUP = 0x03, /**< Rejected: Not Supported */
+};
+
+/** @brief Abstracts memory buffers. */
+struct mqtt_sn_data {
+	const uint8_t *data; /**< Pointer to data. */
+	uint16_t size; /**< Size of data, in bytes. */
+};
+
+/**
+ * @brief Initialize memory buffer from C literal string.
+ *
+ * Use it as follows:
+ *
+ * struct mqtt_sn_data password = MQTT_SN_DATA_STRING_LITERAL("my_pass");
+ *
+ * @param[in] literal Literal string from which to generate mqtt_sn_data object.
+ */
+#define MQTT_SN_DATA_STRING_LITERAL(literal) ((struct mqtt_sn_data){ literal, sizeof(literal) - 1 })
+
+/**
+ * @brief Initialize memory buffer from single bytes.
+ *
+ * Use it as follows:
+ *
+ * struct mqtt_sn_data password = MQTT_SN_DATA_BYTES(0x13, 0x37);
+ */
+#define MQTT_SN_DATA_BYTES(...)                                                                    \
+	((struct mqtt_sn_data){ (uint8_t[]){ __VA_ARGS__ }, sizeof((uint8_t[]){ __VA_ARGS__ }) })
+
+/**
+ * Forward declaration. Library will return pointers to this type only.
+ */
+struct mqtt_sn_msg;
+
+/**
+ * @brief Increase the ref-count of an MQTT-SN message.
+ * 
+ * Use this function in the msg_send callback when you need the associated memory
+ * to remain valid until after the callback returns.
+ * Don't forget to also call mqtt_sn_msg_unref when making use of that functionality.
+ */
+struct mqtt_sn_msg *mqtt_sn_msg_ref(struct mqtt_sn_msg *msg);
+
+/**
+ * @brief Decrease the ref-count of an MQTT-SN message.
+ * 
+ * When the ref-count drops to 0, the message will be de-allocated.
+ */
+void mqtt_sn_msg_unref(struct mqtt_sn_msg *msg);
+
+/**
+ * Get a pointer to the message's content.
+ */
+void mqtt_sn_msg_data(struct mqtt_sn_msg *msg, struct mqtt_sn_data *data);
+
+/**
+ * @brief Callback structure to be used with mqtt_sn_client_init.
+ */
+struct mqtt_sn_callbacks {
+	/** Will be called by the library when it wants to send a message. */
+	int (*msg_send)(struct mqtt_sn_msg *msg, const void *recipient);
+	/** Will be called by the library when it wants to broadcast a message. */
+	int (*msg_broadcast)(struct mqtt_sn_msg *msg, uint8_t radius);
+	/** Will be called by the library when it receives a GWINFO message. */
+	void (*gwinfo)(uint8_t gwid, struct mqtt_sn_data *gw_add);
+	/** Will be called by the library when it receives a PUBLISH message. */
+	void (*publish)(struct mqtt_sn_data *data, enum mqtt_sn_topic_type type, uint16_t tid);
```

Comment:
Again, this is slightly different from our MQTT library, where we only register a single callback, which can handle different event types. I think we could do the same here, to remain consistent.

Reply:
That might make sense, even though this library exposes fewer events than the original MQTT library. For example, I handle stuff like ACKs internally, where the MQTT library leaves that as an exercise to the user.